imap/conversations.c

```diff
@@ -3305,4 +3352,33 @@ EXPORTED int conversations_zero_modseq(struct conversations_state *state)
     return r;
 }
 
+EXPORTED int conversations_lookup_jmapid(struct conversations_state *state,
+                                         const char *jidrep, char guidrep[])
+{
+    char key[CONV_JMAPID_SIZE+2] = "J";
+    size_t datalen = 0;
+    const char *data, *p;
+    int r;
+
+    if (strlen(jidrep) != CONV_JMAPID_SIZE)
```

Comment:
Should this better return an internal error? Or is there any scenario where the caller might legitimately pass a jidrep that doesn't match the expected size?

Reply:
I guess the JMAP code should be checking the id early, and this function can assume that its well-formed.

Should I remove this check?

Reply:
I would rather keep that check but either return CYRUSDB_INTERNAL (and assume that this gets logged properly by the caller), or log the incorrect size here and return NOTFOUND. My point is: if calling that with an invalid size is actually an error by the caller, then this function should not hide that just behind a NOTFOUND.

Reply:
do you have a preference for the log level?

Reply:
I guess I will have to audit every place where _guid_from_id() is called to see if it tries to validate the emailid.

I added the check just to short-circuit doing a DB lookup, because we know its going to fail.

Reply:
[This is just a drive by comment cause something caught my eye in the email notification. I haven't read any of this closely.]

> if calling that with an invalid size is actually an error by the caller [...]

perhaps an assert rather than log-and-return, if it indicates a programmer error or unexpected runtime state

Reply:
Its not really meant to be a check on a programmer error as in some cases jmap_mail.c just passes whatever value it received by the client.  The current comment and code is this:

`    /* sanity check the ID:
       - to prevent strcat() from overrunning a fixed-length buffer
       - to short-circuit an ill-fated DB fetch
    */
    idlen = strspn(jidrep, jmapid_alphabet);
    if (idlen != CONV_JMAPID_SIZE || jidrep[idlen] != '\0') {
        /* we know this J key can't exist */
        return CYRUSDB_NOTFOUND;
    }
`